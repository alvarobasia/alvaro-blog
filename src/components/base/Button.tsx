import React, {useState, ReactElement} from 'react';

interface ButtonProps {
  text: string;
  variant: 'primary' | 'secondary';
  onClick: () => void;
  icon: ReactElement;
}

const Variants = {
  primary: ' p-2 bg-blue-500 hover:bg-blue-700 text-xl transition-all',
  secondary:
    'p-2 border-2 border-blue-500 hover:border-blue-700 hover:bg-blue-700  hover:text-white text-blue-500 text-xl transition-all',
};

const Button: React.FC<ButtonProps> = (props: ButtonProps) => {
  const [isOnMouseOver, setIsOnMouseOver] = useState(false);
  const newIconColor = React.cloneElement(props.icon, {
    color: isOnMouseOver ? 'white' : 'black',
  });
  return (
    <button
      onMouseOver={() => setIsOnMouseOver(true)}
      onMouseLeave={() => setIsOnMouseOver(false)}
      onClick={props.onClick}
      className={
        Variants[props.variant] +
        ' flex flex-row items-center gap-2 rounded group w-40 font-light justify-evenly'
      }
    >
      {newIconColor}
      <p className="group-hover:font-serif group-hover:italic group-hover:text-white">
        {props.text}
      </p>
    </button>
  );
};

export default Button;
