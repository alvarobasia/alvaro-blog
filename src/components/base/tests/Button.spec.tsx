import React from 'react';
import {vi} from 'vitest';
import {MagnifyingGlass} from 'phosphor-react';
import {render, screen} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Button from '../Button';
const mockClick = vi.fn();

describe('Test Button component', () => {
  it('should render Button component', () => {
    const {baseElement} = render(
      <Button
        icon={<MagnifyingGlass />}
        onClick={mockClick}
        text="Teste"
        variant="primary"
      />
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('should call onClick when the user click on button', async () => {
    render(
      <Button
        icon={<MagnifyingGlass />}
        onClick={mockClick}
        text="Teste"
        variant="primary"
      />
    );
    const button = screen.getByRole('button');

    await userEvent.click(button);
    await userEvent.click(button);

    expect(mockClick).toHaveBeenCalledTimes(2);
  });

  it('should render the correct text', () => {
    render(
      <Button
        icon={<MagnifyingGlass />}
        onClick={mockClick}
        text="Teste"
        variant="primary"
      />
    );

    const text = screen.getByText('Teste');

    expect(text).toBeInTheDocument();
  });

  it('should render the correct style in primary variant', () => {
    render(
      <Button
        icon={<MagnifyingGlass />}
        onClick={mockClick}
        text="Teste"
        variant="primary"
      />
    );

    const button = screen.getByRole('button');

    expect(button).toHaveClass(
      'p-2 bg-blue-500 hover:bg-blue-700 text-xl transition-all'
    );

    expect(button).not.toHaveClass(
      'p-2 border-2 border-blue-500 hover:border-blue-700 hover:bg-blue-700  hover:text-white text-blue-500 text-xl transition-all'
    );
  });

  it('should render the correct style in secondary variant', () => {
    render(
      <Button
        icon={<MagnifyingGlass />}
        onClick={mockClick}
        text="Teste"
        variant="secondary"
      />
    );

    const button = screen.getByRole('button');

    expect(button).toHaveClass(
      'p-2 border-2 border-blue-500 hover:border-blue-700 hover:bg-blue-700  hover:text-white text-blue-500 text-xl transition-all'
    );

    expect(button).not.toHaveClass(
      'p-2 bg-blue-500 hover:bg-blue-700 text-xl transition-all'
    );
  });

  it('should render the icon', () => {
    render(
      <Button
        icon={<div data-testid="test" />}
        onClick={mockClick}
        text="Teste"
        variant="secondary"
      />
    );

    const icon = screen.getByTestId('test');

    expect(icon).toBeInTheDocument();
  });
});
