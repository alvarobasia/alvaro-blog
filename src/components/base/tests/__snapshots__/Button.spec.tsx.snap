// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test Button component > should render Button component 1`] = `
<body>
  <div>
    <button
      class=" p-2 bg-blue-500 hover:bg-blue-700 text-xl transition-all flex flex-row items-center gap-2 rounded group w-40 font-light justify-evenly"
    >
      <svg
        fill="black"
        height="1em"
        viewBox="0 0 256 256"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="256"
          width="256"
        />
        <circle
          cx="116"
          cy="116"
          fill="none"
          r="84"
          stroke="black"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
        />
        <line
          fill="none"
          stroke="black"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
          x1="175.4"
          x2="224"
          y1="175.4"
          y2="224"
        />
      </svg>
      <p
        class="group-hover:font-serif group-hover:italic group-hover:text-white"
      >
        Teste
      </p>
    </button>
  </div>
</body>
`;

exports[`Test Button component > should render Button component 2`] = `
<body>
  <div>
    <button
      class=" p-2 bg-blue-500 hover:bg-blue-700 text-xl transition-all flex flex-row items-center gap-2 rounded group w-40 font-light justify-evenly"
    >
      <svg
        fill="black"
        height="1em"
        viewBox="0 0 256 256"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="256"
          width="256"
        />
        <circle
          cx="116"
          cy="116"
          fill="none"
          r="84"
          stroke="black"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
        />
        <line
          fill="none"
          stroke="black"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
          x1="175.4"
          x2="224"
          y1="175.4"
          y2="224"
        />
      </svg>
      <p
        class="group-hover:font-serif group-hover:italic group-hover:text-white"
      >
        Teste
      </p>
    </button>
  </div>
</body>
`;
